src/api/block_builder/input_selection/sender_issuer.rs

```diff
@@ -252,14 +199,183 @@ impl<'a> ClientBlockBuilder<'a> {
+                    if let Output::Alias(alias_output) = &input_signing_data.output {
+                        if alias_id == alias_output.alias_id() {
+                            selected_inputs.push(input_signing_data.clone());
+                            selected_inputs_output_ids.insert(output_id);
+                            // break when we added the required output to the selected_inputs
+                            continue 'addresses_loop;
+                        }
+                    }
+                }
+                return Err(Error::MissingInput(format!(
+                    "missing alias {alias_id} for sender or issuer feature"
+                )));
+            }
+            Address::Nft(nft_address) => {
+                // Check if output is nft address.
+                let nft_id = nft_address.nft_id();
+
+                // check if already selected.
+                for selected_input in selected_inputs.iter() {
+                    if let Output::Nft(nft_output) = &selected_input.output {
+                        if nft_id == nft_output.nft_id() {
+                            continue 'addresses_loop;
+                        }
+                    }
+                }
+
+                // if not found, check currently not selected outputs
+                for input_signing_data in inputs.clone() {
+                    // Skip already added inputs
+                    let output_id = input_signing_data.output_id()?;
+                    if selected_inputs_output_ids.contains(&output_id) {
+                        continue;
+                    }
+                    if let Output::Nft(nft_output) = &input_signing_data.output {
+                        if nft_id == nft_output.nft_id() {
+                            selected_inputs.push(input_signing_data.clone());
+                            selected_inputs_output_ids.insert(output_id);
+                            // break when we added the required output to the selected_inputs
+                            continue 'addresses_loop;
+                        }
+                    }
+                }
+                return Err(Error::MissingInput(format!(
+                    "missing nft {nft_id} for sender or issuer feature"
+                )));
+            }
+        }
+    }
+
+    Ok(())
+}
+
+// Returns required addresses for sender and issuer features that aren't already unlocked with the selected_inputs
+fn get_required_addresses_for_sender_and_issuer(
+    selected_inputs: &[InputSigningData],
+    outputs: &Vec<Output>,
+    current_time: u32,
+) -> crate::Result<HashSet<Address>> {
+    log::debug!("[get_required_addresses_for_sender_and_issuer]");
+
+    let mut all_required_addresses = HashSet::new();
+
+    for output in outputs {
+        if let Some(sender_feature) = output.features().and_then(Features::sender) {
+            // Only add if not already present in the selected inputs.
+            if !selected_inputs.iter().any(|input_data| {
```

Comment:
This looks pretty heavy, nested loop, so maybe we should do a first check if the address is not already in `all_required_addresses` ?